src/java/org/apache/cassandra/service/accord/serializers/TopologySerializers.java

```diff
@@ -159,4 +167,127 @@ public long serializedSize(Topology topology)
             return size;
         }
     };
+
+    public static final UnversionedSerializer<Topology> compactTopology = new UnversionedSerializer<>()
+    {
+        @Override
+        public void serialize(Topology topology, DataOutputPlus out) throws IOException
+        {
+            out.writeLong(topology.epoch());
+            CollectionSerializers.serializeList(topology.staleIds(), out, TopologySerializers.nodeId);
+
+            List<Shard> shards = topology.shards();
+
+            // need to loop twice; once to collect tables/ranges, and another to save shards
+            ImmutableUniqueList<TableId> tables;
+            ImmutableUniqueList<TokenRange> ranges;
+            {
+                ImmutableUniqueList.Builder<TableId> tablesBuilder = ImmutableUniqueList.builder();
+                ImmutableUniqueList.Builder<TokenRange> rangesBuilder = ImmutableUniqueList.builder();
+                for (Shard shard : shards)
+                {
+                    TokenRange range = (TokenRange) shard.range;
+                    tablesBuilder.add(range.table());
+                    rangesBuilder.add(range.withTable(TableId.UNDEFINED));
+                }
+                tables = tablesBuilder.buildAndClear();
+                ranges = rangesBuilder.buildAndClear();
+            }
+
+            CollectionSerializers.serializeList(tables, out, TableId.compactComparableSerializer);
+            CollectionSerializers.serializeList(ranges, out, TokenRange.noTableSerializer);
+
+            out.writeUnsignedVInt32(shards.size());
+            for (Shard shard : shards)
+            {
+                TokenRange range = (TokenRange) shard.range;
+                int tableIdx = tables.indexOf(range.table());
+                int rangeIdx = ranges.indexOf(range.withTable(TableId.UNDEFINED));
+                out.writeUnsignedVInt32(tableIdx);
+                out.writeUnsignedVInt32(rangeIdx);
+
+                CollectionSerializers.serializeList(shard.nodes, out, TopologySerializers.nodeId);
+                BitSet notInFastPath = SortedArrays.toBitSet(shard.nodes, shard.notInFastPath);
+                BitSetSerializer.instance.serialize(notInFastPath, out);
+                BitSet joining = SortedArrays.toBitSet(shard.nodes, shard.joining);
+                BitSetSerializer.instance.serialize(joining, out);
```

Comment:
for murmur partitioner this actually is one of the main savings

Reply:
that's surprising - should't this collection normally be empty?